Netty jms update

isaacrivriv opened the pull request:
for #20132
for #20131

isaacrivriv committed:
First pass with Netty JMS
- Worked on creating constructs/classes based off currently used Channel Framework entites to Netty entities
- Started updating classes to use Netty constructs
- Updated JMS client bundle imports as necessary

isaacrivriv committed:
Added changes for JMS client to work with Netty

isaacrivriv committed:
Added full client functionality
- Added SSL usage
- Fixed missing functionality

isaacrivriv committed:
Updated JMS Server to use Netty

isaacrivriv commented:
#build

dmuelle commented:
I don't see any messages that need ID review in this PR. Adding `ID review not needed` label. If that's not correct let me know. Thanks

posmith commented:
L2 message review not required

isaacrivriv committed:
Updated code for quiesce and timeout functionality
- Switched up inactivity with heartbeat for JMS and cleaned up
- Updated to not read or write on closed channels
- Added quiesce functionality for endpoints
- Switched up netty transport flag to false by default

isaacrivriv committed:
Additional clean up of unused Netty Objects and updated features to pull in private features

isaacrivriv committed:
Revert feature changes due to mismatch of kind and type

isaacrivriv committed:
Setting netty transport default as false

isaacrivriv commented:
#build

joe-chacko commented on dev/com.ibm.ws.messaging.comms.client/src/com/ibm/ws/sib/jfapchannel/framework/NetworkConnectionFactory.java:
```diff
@@ -42,10 +40,11 @@ public interface NetworkConnectionFactory
-     * 
-     * @return an VirtualConnectionFactory
+     * Destroys the connection object stored
+     * @return if connection was successfully destroyed
```
What happens if the connection was not successfully destroyed?
Maybe add an `@throws` line.

isaacrivriv replied:
Added a throws line for this and more clarification. In this case, it's not a specific connection but the object necessary to start up connections for a factory. For chfw there is a virtual connection factory for managing connections which has its own destroy method but for Netty our bootstrap defines how connections are started so the description was not much accurate. Updated to match this.

joe-chacko commented on dev/com.ibm.ws.messaging.comms.client/src/com/ibm/ws/sib/jfapchannel/impl/CommsOutboundChain.java:
```diff
@@ -127,89 +159,130 @@ void unbindSecureFacet(OutboundSecureFacet oldFacet) {
         return tcpProps;
     }
 
+    @Reference(name = "nettyTlsProvider", cardinality = OPTIONAL, policyOption = GREEDY, unbind = "unbindTlsProviderService")
```
This reference should not be dynamic: it belongs in the constructor parameter list.
Bind/unbind are unnecessary complications and extra code.
Also, we should use metatype to ensure we wait for this when it is expected.

isaacrivriv replied:
I also believe in part it should be in the constructor, originally tried it this way but ran into issues because the TLS bundles is loaded with the SSL feature used for creating Netty SSL configurations. So if a server is configured with just unencrypted traffic, the server will never call the constructor parameter list for the bundle and nothing will ever load. Maybe we can figure something out that'd be better for this, I'll DM you.

joe-chacko commented on dev/com.ibm.ws.messaging.comms.client/src/com/ibm/ws/sib/jfapchannel/impl/CommsOutboundChain.java:
```diff
@@ -91,12 +116,19 @@ public CommsOutboundChain(
 
         isSecureChain = MetatypeUtils.parseBoolean(OUTBOUND_CHAIN_CONFIG_ALIAS, "useSSL", properties.get("useSSL"), false);
+        
+        useNettyTransport = ProductInfo.getBetaEdition() ? 
```
A trigraph operator does not make this clearer. Use 
`Product.getBetaEdition() && MetatypeUtils.parseBoolean(OUTBOUND_CHAIN_CONFIG_ALIAS, "useNettyTransport", properties.get("useNettyTransport"), true)` instead.

isaacrivriv replied:
Sounds good, I'll make this change.

joe-chacko commented on dev/com.ibm.ws.messaging.comms.client/src/com/ibm/ws/sib/jfapchannel/impl/CommsOutboundChain.java:
```diff
@@ -127,89 +159,130 @@ void unbindSecureFacet(OutboundSecureFacet oldFacet) {
-        try {
-            ChannelFramework cfw = commsClientService.getChannelFramework();
-            cfw.registerFactory("JFapChannelOutbound", JFapChannelFactory.class);
+        if(!useNetty()) {
```
Avoid using `if (!condition) {} else {}` — flip the condition and swap the if/else blocks. 
This should be a refactoring in your IDE.

joe-chacko commented on dev/com.ibm.ws.messaging.comms.client/src/com/ibm/ws/sib/jfapchannel/impl/CommsOutboundChain.java:
```diff
@@ -127,89 +159,130 @@ void unbindSecureFacet(OutboundSecureFacet oldFacet) {
+            }
+        }else {
+        	// Use Netty Framework for transport
+            chainList.put(chainName, this);
```
How is this different from the similar looking code in the if block?
If they are meant to be the same, move it outside the else block.

isaacrivriv replied:
In this case they are meant to be the same but can't move the code out of the if/else block to be common because of the try catch block while setting up channelfw which would not add to the chainlist if `ChannelException|ChainException` occurs

joe-chacko commented on dev/com.ibm.ws.messaging.comms.client/src/com/ibm/ws/sib/jfapchannel/impl/Connection.java:
```diff
@@ -634,7 +683,11 @@ private long send(JFapByteBuffer buffer,
             // Proddle the write callback to get it to leap into action and start
             // sending data from the priority queue.
-            writeCompletedCallback.proddle();
+            // TODO: Question ..this makes me think about a Proddable or WriteCompletedCallback interface type abstraction (not clear if it would be better but maybe)
```
Definitely create the interface and get rid of this checking and duplication.

isaacrivriv replied:
Will work on it.

joe-chacko reviewed (changes requested):
See comments and DMs for details.

isaacrivriv committed:
Fixed review comments
- Updated metatype parsing for clarity
- Created common class for read and write callbacks for functionality
- Cleaned up and clarified comments

isaacrivriv committed:
Test using config processing for tls bundle load up

isaacrivriv commented:
#libby

isaacrivriv commented:
#build

isaacrivriv commented:
#build

isaacrivriv commented:
#libby

joe-chacko committed:
Use metatype alias for netty TLS provider

joe-chacko committed:
Set cardinality, default, and ibm:final on nettyTlsRef AD

joe-chacko commented:
#build #run-libby-bot

isaacrivriv committed:
Updated netty tls bundle to remove localization in metatype and added
missing entry for netty ssl

isaacrivriv committed:
Restore metatype default to false

isaacrivriv commented:
#libby

isaacrivriv commented:
#build

isaacrivriv commented:
#build

joe-chacko commented on dev/com.ibm.ws.messaging.comms.client/src/com/ibm/ws/sib/jfapchannel/impl/CommsOutboundChain.java:
```diff
@@ -304,4 +391,29 @@ private void removeChannel(String channelName) {
+	 * Get the SSL Options for this Chain
+	 * @return the OutboundSecureFacet wrapper of SSL Options for the chain
+	 */
+	public OutboundSecureFacet getSecureFacet() {
```
Do we need to expose the OutboundSecureFacet here, or could we just return the SSL Options?
Returning the facet is tighter coupling because it is an internal object, whereas the SSL Options comes from another component's SPI.

isaacrivriv replied:
Sounds good, made this change

joe-chacko commented on dev/com.ibm.ws.messaging.comms.client/src/com/ibm/ws/sib/jfapchannel/impl/Connection.java:
```diff
@@ -108,9 +110,9 @@ public abstract class Connection implements ConnectionInterface
 
     // Callbacks notified when a read or write completes for this connection
-    private ConnectionWriteCompletedCallback writeCompletedCallback = null;
+    private BaseConnectionWriteCallback writeCompletedCallback = null;
```
It looks like this can be final since it is only ever assigned inside a constructor.

joe-chacko commented on dev/com.ibm.ws.messaging.comms.server/src/com/ibm/ws/jfap/inbound/channel/CommsServerServiceFacade.java:
```diff
@@ -62,27 +64,35 @@
-    private final CommsInboundChain inboundBasicChain = new CommsInboundChain(this, false);
-    private final CommsInboundChain inboundSecureChain = new CommsInboundChain(this, true);
+    
+    private InboundChain inboundBasicChain;
```
Can we make this final?

joe-chacko commented on dev/com.ibm.ws.messaging.comms.server/src/com/ibm/ws/jfap/inbound/channel/CommsServerServiceFacade.java:
```diff
@@ -62,27 +64,35 @@
-    private final CommsInboundChain inboundSecureChain = new CommsInboundChain(this, true);
+    
+    private InboundChain inboundBasicChain;
+    private InboundChain inboundSecureChain;
```
Can we make this final?

joe-chacko commented on dev/com.ibm.ws.messaging.comms.server/src/com/ibm/ws/jfap/inbound/channel/InboundChain.java:
```diff
@@ -0,0 +1,20 @@
+
+public interface InboundChain {
+	
+	public void enable(boolean enabled);
```
don't need the word public here — it's implicit in an interface that all its members are public

isaacrivriv replied:
Did not know this, I will make the change!

isaacrivriv committed:
Added review changes
- Updated to use final variables where possible
- Updated to remove unnecessary instanceof/casting
- Updated to fix other comments

isaacrivriv commented:
#libby

isaacrivriv commented:
#build

joe-chacko reviewed (approved):
Good work

isaacrivriv commented:
#build

isaacrivriv commented:
Errors in build are related to com.ibm.ws.security.utility_fat FAT tests nothing JMS related. Should be safe to merge in
